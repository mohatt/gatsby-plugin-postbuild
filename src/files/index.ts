export { default as FileType } from './type'
export { default as File } from './base'
export { default as FileGeneric } from './generic'
export { default as FileHtml } from './html'
