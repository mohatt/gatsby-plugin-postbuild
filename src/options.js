export const defaults = {
  enabled: true,
  report: true,
  reportConsole: true,
  purgecss: {
    allowSymbols: false,
    rejected: true,
    fontFace: false,
    keyframes: false,
    variables: false
  }
}

export function schema (Joi) {
  return Joi.object({
    enabled: Joi.boolean()
      .description('Whether to run the purgecss task or not.'),
    report: Joi.boolean()
      .description('Write a "postbuild.log.json" file in "public" with all the changes made.'),
    reportConsole: Joi.boolean()
      .description('Print a summary report during build with all the changes made.'),
    purgecss: Joi.object({
      allowSymbols: Joi.boolean()
        .description('Allow CSS selectors to contain symbols (e.g TailwindCSS).'),
      rejected: Joi.boolean()
        .description('Write a log file in "public" with the rejected selectors for every file.'),
      defaultExtractor: Joi.function()
        .description('A custom PurgeCSS extractor to be used instead of the default one.'),
      safelist: Joi.alternatives().try(
        Joi.object({}).unknown(true),
        Joi.array().items(Joi.string())
      ).description('Selectors that are safe to leave in the final CSS.'),
      blocklist: Joi.array().items(Joi.string())
        .description('Selectors that are blocked from appearing in the final CSS.'),
      fontFace: Joi.boolean()
        .description('Remove any unused @font-face rules in your css.'),
      keyframes: Joi.boolean()
        .description('Remove any unused @keyframes rules in your css.'),
      variables: Joi.boolean()
        .description('Remove any unused variables in your css.')
    })
  })
}
