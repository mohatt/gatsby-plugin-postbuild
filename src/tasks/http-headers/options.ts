import type { ITaskApiOptions, ITaskOptions } from '@postbuild'
import { PROVIDER_TYPES, ProviderType } from './lib/providers'
import type Link from './lib/link'
import type Meta from './lib/meta'

/**
 * Header value is either a string for single-value headers or
 *  an array of strings for multi-value headers
 *
 *  Example of a single-value header:
 *   {
 *     'cache-control': 'public, max-age=31536000, immutable'
 *   }
 *
 *  Examples of a multi-value header:
 *   {
 *    'link': [
 *      '</script.js>; rel=preload; as=script',
 *      '</image.png>; rel=preload; as=image'
 *    ]
 *   }
 */
export interface IHeadersMap {
  [name: string]: string | string[]
}

/**
 * Map of paths to path headers
 */
export interface IPathHeadersMap {
  [path: string]: IHeadersMap
}

/**
 * Task options interface
 */
export interface IHttpHeadersTaskOptions extends ITaskOptions {
  provider: ProviderType
  headers: IPathHeadersMap
  security: boolean
  caching: boolean
  transformPathLinks: (links: Link[], path: string) => Link[]
  removeLinkTags: boolean
  transformPathMetas: (metas: Meta[], path: string) => Meta[]
  removeMetaTags: boolean
}

/**
 * Options API exports
 */
export const options: ITaskApiOptions<IHttpHeadersTaskOptions> = {
  defaults: {
    enabled: false,
    ignore: [],
    provider: 'netlify',
    headers: {},
    security: true,
    caching: true,
    transformPathLinks: (links) => links,
    removeLinkTags: true,
    transformPathMetas: (metas) => metas,
    removeMetaTags: true,
  },
  schema: (joi) => {
    const nonEmptyString = joi.string().trim().required()
    return joi.object({
      provider: joi
        .string()
        .valid(...PROVIDER_TYPES)
        .description('Headers file provider.'),
      headers: joi
        .object()
        .pattern(
          nonEmptyString,
          joi
            .object()
            .pattern(
              joi.string().pattern(/^[\w-]+$/),
              joi.alternatives().try(nonEmptyString, joi.array().items(nonEmptyString)),
            ),
        )
        .description('Headers to merge with the generated headers.'),
      security: joi.boolean().description('Adds useful security headers.'),
      caching: joi.boolean().description('Adds useful caching headers to immutable asset paths.'),
      transformPathLinks: joi
        .function()
        .maxArity(2)
        .description('Callback for manipulating links under each path.'),
      removeLinkTags: joi
        .boolean()
        .description('Removes the HTML link tags processed by the task.'),
      transformPathMetas: joi
        .function()
        .maxArity(2)
        .description('Callback for manipulating metas under each path.'),
      removeMetaTags: joi
        .boolean()
        .description('Removes the HTML meta tags processed by the task.'),
    })
  },
}

declare module 'gatsby-plugin-postbuild' {
  interface IOptions {
    'http-headers': IHttpHeadersTaskOptions
  }
}
