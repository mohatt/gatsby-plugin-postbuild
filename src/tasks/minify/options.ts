import type { MinifyOptions as TerserOptions } from 'terser'
import type { ITaskApiOptions, ITaskOptions } from '@postbuild'

/**
 * Task options interface
 */
export interface IMinifyTaskOptions extends ITaskOptions {
  script: TerserOptions | boolean
  style: [string, object] | object | Function | boolean
}

/**
 * Options API exports
 */
export const options: ITaskApiOptions<IMinifyTaskOptions> = {
  defaults: {
    enabled: false,
    ignore: [],
    script: {},
    style: ['default', {}],
  },
  schema: (joi) => {
    return joi.object({
      script: joi
        .alternatives()
        .try(joi.boolean(), joi.object().unknown(true))
        .description('Options passed to javascript minifier.'),
      style: joi
        .alternatives()
        .try(
          joi.boolean(),
          joi.object(),
          joi.array().items(joi.string(), joi.object().unknown(true)),
          joi.function().maxArity(0),
        )
        .description('Options passed to css minifier.'),
    })
  },
}

declare module 'gatsby-plugin-postbuild' {
  interface IOptions {
    minify: IMinifyTaskOptions
  }
}
