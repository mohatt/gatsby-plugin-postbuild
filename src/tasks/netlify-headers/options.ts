import type { ITaskApiOptions, ITaskOptions } from '~/tasks'

/**
 * Task options interface
 */
export type IOptions = ITaskOptions & {
  headers: {
    [path: string]: string[]
  }
  security: boolean
  caching: boolean
  cachingAssetTypes: string[]
  removeLinkTags: boolean
}

/**
 * Options API exports
 */
export const options: ITaskApiOptions<IOptions> = {
  defaults: {
    enabled: true,
    ignore: [],
    headers: {},
    security: true,
    caching: true,
    cachingAssetTypes: ['image', 'script', 'style', 'font'],
    removeLinkTags: true
  },
  // @todo write desc
  schema: (joi) => {
    return joi.object({
      headers: joi.object()
        .pattern(joi.string(), joi.array().items(joi.string()))
        .description('@todo'),
      security: joi.boolean()
        .description('@todo'),
      caching: joi.boolean()
        .description('@todo'),
      cachingAssetTypes: joi.array()
        .items(joi.string())
        .description('@todo'),
      removeLinkTags: joi.boolean()
        .description('@todo')
    })
  }
}
