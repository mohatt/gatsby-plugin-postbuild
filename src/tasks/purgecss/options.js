export const defaults = {
  // Options consumend by the task
  enabled: true,
  report: true,
  reportConsole: true,
  tailwind: false,
  // Options consumend by both the task and PurgeCSS
  rejected: true
  // Options consumend by PurgeCss
  // We leave that to PurgeCss
}

export function schema (Joi) {
  return Joi.object({
    enabled: Joi.boolean()
      .description('Whether to run the purgecss task or not.'),
    report: Joi.boolean()
      .description('Write a "purgecss.log.json" file in "public" with all the changes made.'),
    reportConsole: Joi.boolean()
      .description('Print a summary report during build with all the changes made.'),
    tailwind: Joi.boolean()
      .description('Use custom extractor for tailwind.'),
    rejected: Joi.boolean()
      .description('Write a log file in "public" with the rejected selectors for every file.'),
    defaultExtractor: Joi.function()
      .description('A custom PurgeCSS extractor to be used instead of the default one.'),
    extractors: Joi.array().items(Joi.object({}).unknown(true))
      .description('A list of custom PurgeCSS extractors to be used for certain file types.'),
    safelist: Joi.alternatives().try(
      Joi.object({}).unknown(true),
      Joi.array().items(Joi.string())
    ).description('Selectors that are safe to leave in the final CSS.'),
    blocklist: Joi.array().items(Joi.string())
      .description('Selectors that are blocked from appearing in the final CSS.'),
    fontFace: Joi.boolean()
      .description('Remove any unused @font-face rules in your css.'),
    keyframes: Joi.boolean()
      .description('Remove any unused @keyframes rules in your css.'),
    variables: Joi.boolean()
      .description('Remove any unused variables in your css.')
  })
}
