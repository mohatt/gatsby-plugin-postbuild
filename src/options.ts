import { GatsbyJoi } from './gatsby'
import { defaultOptions } from 'purgecss'
export type PurgeCSSOptions = typeof defaultOptions

/**
 * Plugin options interface
 */
export interface IPluginOptions {
  enabled: boolean
  report: boolean
  reportConsole: boolean
  allowSymbols: boolean
  ignoreFiles: {
    webpack: string[]
    pages: string[]
    css: string[]
    js: string[]
  }
  purgecss: Pick<PurgeCSSOptions, 'defaultExtractor'|'rejected'|'fontFace'|'keyframes'|'variables'|'blocklist'|'safelist'>
}

/**
 * Default values for plugin options
 *  we are merging purgecss defaults as well while
 *  overridding some values
 */
export const defaults: IPluginOptions = {
  enabled: true,
  report: true,
  reportConsole: true,
  ignoreFiles: {
    webpack: ['app', 'polyfill'],
    pages: [],
    css: [],
    js: []
  },
  allowSymbols: false,
  purgecss: {
    ...defaultOptions,
    rejected: true
  }
}

/**
 * Plugin options schema
 */
export function schema (Joi: GatsbyJoi): GatsbyJoi {
  return Joi.object({
    enabled: Joi.boolean()
      .description('Whether to run the purgecss task or not.'),
    report: Joi.boolean()
      .description('Write a "postbuild.log.json" file in "/public" with all the changes made.'),
    reportConsole: Joi.boolean()
      .description('Print a summary report during build with all the changes made.'),
    ignoreFiles: Joi.object({
      webpack: Joi.array().items(Joi.string())
        .description('Webpack chunck names to ignore while purging CSS.'),
      pages: Joi.array().items(Joi.string())
        .description('Pages to exclude from optimiztion.'),
      css: Joi.array().items(Joi.string())
        .description('Webpack chunck names to ignore while purging CSS.'),
      js: Joi.array().items(Joi.string())
        .description('Webpack chunck names to ignore while purging CSS.')
    }).description('CSS/JavaScript files to ignore during optimization. File names should be relative to "/public" directory'),
    allowSymbols: Joi.boolean()
      .description('Sets a custom PurgeCSS extractor that allows CSS selectors to contain symbols (e.g TailwindCSS).'),
    purgecss: Joi.object({
      rejected: Joi.boolean()
        .description('Write a log file in "/public" with the rejected selectors for every file.'),
      defaultExtractor: Joi.function()
        .description('A custom PurgeCSS extractor to be used instead of the default one.'),
      safelist: Joi.alternatives().try(
        Joi.object({}).unknown(true),
        Joi.array().items(Joi.string())
      ).description('Selectors that are safe to leave in the final CSS.'),
      blocklist: Joi.array().items(Joi.string())
        .description('Selectors that are blocked from appearing in the final CSS.'),
      fontFace: Joi.boolean()
        .description('Remove any unused @font-face rules in your css.'),
      keyframes: Joi.boolean()
        .description('Remove any unused @keyframes rules in your css.'),
      variables: Joi.boolean()
        .description('Remove any unused variables in your css.')
    })
  })
}
