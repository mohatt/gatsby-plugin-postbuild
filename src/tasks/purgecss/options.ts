import _ from 'lodash'
import { defaultOptions as purgecssDefaults } from 'purgecss'
import type { ITaskApiOptions, ITaskOptions } from '~/tasks'
type PurgecssDefaults = typeof purgecssDefaults

/**
 * Extracts union type from array values
 */
type ArrayUnion<T extends readonly unknown[]> = T extends ReadonlyArray<infer A> ? A : never

/**
 * Options imported from PurgeCss
 */
export const purgecssImportedOptions = [
  'defaultExtractor',
  'safelist',
  'blocklist',
  'fontFace',
  'keyframes',
  'variables'
] as const

/**
 * Task options interface
 */
export type IOptions = ITaskOptions & {
  reportRejected: boolean
  allowSymbols: boolean
  ignoreAssets: {
    webpack: string[]
    css: string[]
    js: string[]
  }
} & Pick<PurgecssDefaults, ArrayUnion<typeof purgecssImportedOptions>>

/**
 * All options accepted by purgecss lib
 */
export type PurgecssOptions = Omit<Partial<PurgecssDefaults>, 'content'|'css'> & Pick<PurgecssDefaults, 'content'|'css'>

/**
 * Options API exports
 */
export const options: ITaskApiOptions<IOptions> = {
  defaults: {
    enabled: true,
    ignore: [],
    ignoreAssets: {
      webpack: ['app', 'polyfill'],
      css: [],
      js: []
    },
    reportRejected: true,
    allowSymbols: false,
    ..._.pick(purgecssDefaults, purgecssImportedOptions)
  },
  schema: (joi) => {
    return joi.object({
      reportRejected: joi.boolean()
        .description('Write a log file in `/public` with the rejected selectors for every file.'),
      ignoreAssets: joi.object({
        webpack: joi.array().items(joi.string())
          .description('Webpack chunck names to ignore while purging CSS.'),
        css: joi.array().items(joi.string())
          .description('CSS files to exclude from processing.'),
        js: joi.array().items(joi.string())
          .description('JavaScript files to ignore while purging CSS.')
      }).description('CSS/JavaScript files to ignore during optimization. File names should be relative to `/public` directory'),
      allowSymbols: joi.boolean()
        .description('Sets a custom PurgeCSS extractor that allows CSS selectors to contain symbols (e.g TailwindCSS).'),
      defaultExtractor: joi.function()
        .description('A custom PurgeCSS extractor to be used instead of the default one.'),
      safelist: joi.alternatives().try(
        joi.object({}).unknown(true),
        joi.array().items(joi.string())
      ).description('Selectors that are safe to leave in the final CSS.'),
      blocklist: joi.array().items(joi.string())
        .description('Selectors that are blocked from appearing in the final CSS.'),
      fontFace: joi.boolean()
        .description('Remove any unused `@font-face` rules in your css.'),
      keyframes: joi.boolean()
        .description('Remove any unused `@keyframes` rules in your css.'),
      variables: joi.boolean()
        .description('Remove any unused variables in your css.')
    })
  }
}
