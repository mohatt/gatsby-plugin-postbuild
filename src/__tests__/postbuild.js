import { testPluginOptionsSchema } from 'gatsby-plugin-utils'
import optionsFixtures from '#/__fixtures__/options'
import Postbuild from '~/postbuild'
import { File } from '~/files'
import { DEFAULTS } from '~/options'

// Async mock fn
jest.fnAsync = v => jest.fn().mockImplementation(() => Promise.resolve(v))

describe('init', () => {
  let postbuild
  const tasks = {
    register: jest.fn(),
    getOptionsSchemas: joi => ({
      foo: joi.object({ bar: joi.string() }),
      lorem: joi.object({ ipsum: joi.boolean() })
    })
  }

  test('options defaults', () => {
    expect(() => {
      postbuild = new Postbuild(tasks, {})
      postbuild.init(['foo', 'lorem'])
    }).not.toThrow()
    expect(postbuild.options).toMatchObject(DEFAULTS)
    expect(tasks.register.mock.calls).toMatchSnapshot()
  })

  describe('options schema', () => {
    const schema = ({ Joi }) => postbuild.getOptionsSchemas(Joi)
    optionsFixtures.push({
      title: 'should validate default options',
      options: DEFAULTS
    })
    for (const { title, options } of optionsFixtures) {
      test(title, () => testPluginOptionsSchema(schema, options)
        .then(({ isValid, errors }) => {
          expect(isValid).toMatchSnapshot()
          expect(errors).toMatchSnapshot()
        })
      )
    }
  })
})

describe('bootstrap', () => {
  const tasks = {
    register: jest.fn(),
    setOptions: jest.fn(),
    run: jest.fnAsync(),
    getActiveTasks: () => [{}]
  }
  const filesystem = {
    setRoot: jest.fn()
  }
  const gatsby = {
    store: {
      getState: () => ({
        program: { directory: '/foo/bar' }
      })
    },
    reporter: {
      setErrorMap: jest.fn()
    }
  }

  const cases = [
    {
      title: 'correctly loads with defaults',
      options: {}
    },
    {
      title: 'correctly loads with user options',
      options: {
        events: { foo: { bar: () => '' } }
      }
    }
  ]

  beforeEach(() => {
    tasks.register.mockClear()
    tasks.setOptions.mockClear()
    tasks.run.mockClear()
    filesystem.setRoot.mockClear()
  })
  for (const { title, options } of cases) {
    test(title, async () => {
      const postbuild = new Postbuild(tasks, filesystem)
      await expect(postbuild.bootstrap(gatsby, options)).resolves.toBe(undefined)
      expect(postbuild.options).toMatchObject(Object.keys(options).length ? options : DEFAULTS)
      expect(filesystem.setRoot).toBeCalledWith('/foo/bar/public')
      expect(tasks.register.mock.calls).toMatchSnapshot()
      expect(tasks.setOptions).toBeCalledTimes(1)
      expect(tasks.run.mock.calls[0])
        .toMatchObject(
          ['on', 'bootstrap', { filesystem, gatsby }]
        )
    })
  }

  test('incorrectly ignores running when no active tasks', async () => {
    tasks.getActiveTasks = () => []
    const postbuild = new Postbuild(tasks, filesystem)
    await expect(postbuild.bootstrap(gatsby, {})).resolves.toBe(undefined)
    expect(postbuild.options.enabled).toBe(false)
    expect(tasks.run.mock.calls.length).toBe(0)
  })
})

describe('run', () => {
  File.factory = jest.fn()
  const tasks = {
    getFilenames: jest.fnAsync({
      foo: ['file1.foo', 'file2.foo']
    }),
    run: jest.fnAsync()
  }
  const setStatus = jest.fn()
  const filesystem = {
    create: jest.fnAsync(),
    reporter: {
      getReports: () => ['report1', 'report2'],
      getTotalSaved: () => [5, 5]
    }
  }
  const gatsby = {}

  beforeEach(() => {
    File.factory.mockClear()
    setStatus.mockClear()
    tasks.run.mockClear()
    filesystem.create.mockClear()
  })

  const runTest = (options = {}) => {
    const pb = new Postbuild(tasks, filesystem)
    pb.options = { ...pb.options, ...options }
    pb.process = jest.fn().mockImplementation((a, b, tick) => {
      [0, 0].forEach(() => {
        tick('read'); tick('process'); tick('write')
      })
    })
    return expect(pb.run(gatsby, setStatus)).resolves.toBe(undefined).then(() => {
      expect(File.factory.mock.calls)
        .toMatchObject([
          ['foo', 'file1.foo', filesystem, tasks, gatsby],
          ['foo', 'file2.foo', filesystem, tasks, gatsby]
        ])
      expect(tasks.run.mock.calls)
        .toMatchObject([
          ['on', 'postbuild', { filesystem, gatsby }],
          ['foo', 'configure', { filesystem, gatsby, config: pb.options.processing }],
          ['on', 'shutdown', { filesystem, gatsby }]
        ])
      expect(pb.process).toBeCalledTimes(1)
      return pb
    })
  }

  test('correctly runs with default options', () => runTest().then(() => {
    expect(setStatus.mock.calls).toMatchSnapshot()
  }))

  test('correctly runs with custom processing options', () => runTest({
    processing: { strategy: 'parallel', concurrency: 3 }
  }).then(pb => {
    expect(pb.process.mock.calls[0][1]).toMatchObject({
      strategy: 'parallel',
      concurrency: 3
    })
  }))

  test('correctly runs with custom extension options', () => runTest({
    extensions: {
      foo: { strategy: 'parallel', concurrency: 2 }
    }
  }).then(pb => {
    expect(pb.process.mock.calls[0][1]).toMatchObject({
      strategy: 'parallel',
      concurrency: 2
    })
  }))

  test('correctly writes log file', () => runTest({ reporting: { log: true } }).then(() => {
    expect(filesystem.create.mock.calls).toMatchSnapshot()
  }))

  test('correctly ignores writing log file', () => runTest({ reporting: { log: false } }).then(() => {
    expect(filesystem.create.mock.calls).toMatchSnapshot()
  }))
})

describe('process', () => {
  const tick = jest.fn()
  const file = {
    read: jest.fnAsync(),
    process: jest.fnAsync(),
    write: jest.fnAsync()
  }
  const files = [file, file, file, file]

  beforeEach(() => {
    tick.mockClear()
    file.read.mockClear()
    file.process.mockClear()
    file.write.mockClear()
  })

  const cases = [
    { strategy: 'parallel', concurrency: 1 },
    { strategy: 'parallel', concurrency: 2 },
    { strategy: 'parallel', concurrency: 3 },
    { strategy: 'sequential', concurrency: 1 },
    { strategy: 'sequential', concurrency: 2 },
    { strategy: 'sequential', concurrency: 3 }
  ]
  for (const c of cases) {
    test(`correctly runs with s:${c.strategy} c:${c.concurrency}`, async () => {
      const postbuild = new Postbuild({}, {})
      postbuild.files.foo = files.concat()
      await expect(postbuild.process('foo', c, tick)).resolves.toBe(undefined)
      expect(file.read.mock.calls.length).toBe(4)
      expect(file.process.mock.calls.length).toBe(4)
      expect(file.write.mock.calls.length).toBe(4)
      expect(tick.mock.calls.toString()).toMatchSnapshot()
    })
  }
})
