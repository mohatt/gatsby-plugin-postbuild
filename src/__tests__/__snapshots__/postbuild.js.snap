// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init options defaults 1`] = `
Array [
  Array [
    "foo",
  ],
  Array [
    "lorem",
  ],
]
`;

exports[`init options schema should invalidate empty options 1`] = `false`;

exports[`init options schema should invalidate empty options 2`] = `
Array [
  "\\"ignore[0]\\" is not allowed to be empty",
  "\\"extensions.txt.concurrency\\" must be greater than or equal to 1",
  "\\"extensions.txt.strategy\\" must be one of [steps, parallel]",
  "\\"extensions.txt.strategy\\" is not allowed to be empty",
  "\\"defaultStrategy\\" must be one of [steps, parallel]",
  "\\"defaultStrategy\\" is not allowed to be empty",
  "\\"defaultConcurrency\\" must be greater than or equal to 1",
]
`;

exports[`init options schema should invalidate invalid options 1`] = `false`;

exports[`init options schema should invalidate invalid options 2`] = `
Array [
  "\\"events.on.load\\" is not allowed",
  "\\"events.html.content\\" is not allowed",
  "\\"events.txt.parse\\" is not allowed",
  "\\"extensions.txt.concurrency\\" must be greater than or equal to 1",
  "\\"extensions.txt.strategy\\" must be one of [steps, parallel]",
  "\\"defaultStrategy\\" must be one of [steps, parallel]",
  "\\"defaultConcurrency\\" must be greater than or equal to 1",
]
`;

exports[`init options schema should invalidate invalid tasks options 1`] = `false`;

exports[`init options schema should invalidate invalid tasks options 2`] = `
Array [
  "\\"foo.bar\\" must be a string",
  "\\"lorem\\" must be of type object",
]
`;

exports[`init options schema should invalidate options with incorrect data types 1`] = `false`;

exports[`init options schema should invalidate options with incorrect data types 2`] = `
Array [
  "\\"enabled\\" must be a boolean",
  "\\"report\\" must be a boolean",
  "\\"ignore\\" must be an array",
  "\\"events.on.bootstrap\\" must be of type function",
  "\\"events.html.parse\\" must be of type function",
  "\\"events.foo\\" must be of type object",
  "\\"events.txt.content\\" must be of type function",
  "\\"extensions.foo\\" must be of type object",
  "\\"extensions.txt.concurrency\\" must be a number",
  "\\"extensions.txt.strategy\\" must be one of [steps, parallel]",
  "\\"extensions.txt.strategy\\" must be a string",
  "\\"defaultStrategy\\" must be one of [steps, parallel]",
  "\\"defaultStrategy\\" must be a string",
  "\\"defaultConcurrency\\" must be a number",
]
`;

exports[`init options schema should invalidate unknown options 1`] = `false`;

exports[`init options schema should invalidate unknown options 2`] = `
Array [
  "\\"unknown\\" is not allowed",
  "\\"invalid\\" is not allowed",
]
`;

exports[`init options schema should validate correct options 1`] = `true`;

exports[`init options schema should validate correct options 2`] = `Array []`;

exports[`init options schema should validate correct tasks options 1`] = `true`;

exports[`init options schema should validate correct tasks options 2`] = `Array []`;

exports[`init options schema should validate default options 1`] = `true`;

exports[`init options schema should validate default options 2`] = `Array []`;

exports[`run correctly ignores writing report file 1`] = `Array []`;

exports[`run correctly runs mixed strats 1`] = `
Array [
  Array [
    "Loaded 1/4 Processed 0/4 Wrote 0/4",
  ],
  Array [
    "Loaded 2/4 Processed 0/4 Wrote 0/4",
  ],
  Array [
    "Loaded 2/4 Processed 1/4 Wrote 0/4",
  ],
  Array [
    "Loaded 2/4 Processed 2/4 Wrote 0/4",
  ],
  Array [
    "Loaded 2/4 Processed 2/4 Wrote 1/4",
  ],
  Array [
    "Loaded 2/4 Processed 2/4 Wrote 2/4",
  ],
  Array [
    "Loaded 3/4 Processed 2/4 Wrote 2/4",
  ],
  Array [
    "Loaded 3/4 Processed 3/4 Wrote 2/4",
  ],
  Array [
    "Loaded 3/4 Processed 3/4 Wrote 3/4",
  ],
  Array [
    "Loaded 4/4 Processed 3/4 Wrote 3/4",
  ],
  Array [
    "Loaded 4/4 Processed 4/4 Wrote 3/4",
  ],
  Array [
    "Loaded 4/4 Processed 4/4 Wrote 4/4",
  ],
  Array [
    "Done processing 4 files saving a total of 5",
  ],
]
`;

exports[`run correctly runs parallel strat 1`] = `
Array [
  Array [
    "Loaded 1/4 Processed 0/4 Wrote 0/4",
  ],
  Array [
    "Loaded 1/4 Processed 1/4 Wrote 0/4",
  ],
  Array [
    "Loaded 1/4 Processed 1/4 Wrote 1/4",
  ],
  Array [
    "Loaded 2/4 Processed 1/4 Wrote 1/4",
  ],
  Array [
    "Loaded 2/4 Processed 2/4 Wrote 1/4",
  ],
  Array [
    "Loaded 2/4 Processed 2/4 Wrote 2/4",
  ],
  Array [
    "Loaded 3/4 Processed 2/4 Wrote 2/4",
  ],
  Array [
    "Loaded 3/4 Processed 3/4 Wrote 2/4",
  ],
  Array [
    "Loaded 3/4 Processed 3/4 Wrote 3/4",
  ],
  Array [
    "Loaded 4/4 Processed 3/4 Wrote 3/4",
  ],
  Array [
    "Loaded 4/4 Processed 4/4 Wrote 3/4",
  ],
  Array [
    "Loaded 4/4 Processed 4/4 Wrote 4/4",
  ],
  Array [
    "Done processing 4 files saving a total of 5",
  ],
]
`;

exports[`run correctly runs steps strat 1`] = `
Array [
  Array [
    "Loaded 1/4 Processed 0/4 Wrote 0/4",
  ],
  Array [
    "Loaded 2/4 Processed 0/4 Wrote 0/4",
  ],
  Array [
    "Loaded 2/4 Processed 1/4 Wrote 0/4",
  ],
  Array [
    "Loaded 2/4 Processed 2/4 Wrote 0/4",
  ],
  Array [
    "Loaded 2/4 Processed 2/4 Wrote 1/4",
  ],
  Array [
    "Loaded 2/4 Processed 2/4 Wrote 2/4",
  ],
  Array [
    "Loaded 3/4 Processed 2/4 Wrote 2/4",
  ],
  Array [
    "Loaded 4/4 Processed 2/4 Wrote 2/4",
  ],
  Array [
    "Loaded 4/4 Processed 3/4 Wrote 2/4",
  ],
  Array [
    "Loaded 4/4 Processed 4/4 Wrote 2/4",
  ],
  Array [
    "Loaded 4/4 Processed 4/4 Wrote 3/4",
  ],
  Array [
    "Loaded 4/4 Processed 4/4 Wrote 4/4",
  ],
  Array [
    "Done processing 4 files saving a total of 5",
  ],
]
`;

exports[`run correctly writes report file 1`] = `
Array [
  Array [
    "postbuild.log.json",
    "[
  \\"report1\\",
  \\"report2\\"
]",
  ],
]
`;
